src/sentry/dynamic_sampling/models/utils.py

```diff
@@ -12,47 +11,177 @@ class DSElement:
+    sorted_elements = sorted(classes, key=lambda x: (x.count, x.id), reverse=True)
+    return _adjust_sample_rates_full(sorted_elements, rate, intensity, min_budget)
+
+
+def _adjust_sample_rates_full(
+    classes: List[DSElement], rate: float, intensity: float, min_budget: Optional[float] = None
+) -> Tuple[List[DSElement], float]:
+    """
+    Tries to calculate rates that brings all counts close to the ideal count
+
+    The intensity controls how close, 0 intensity leaves the items unchanged, 1 brings the items to the
+    ideal count ( or rate=1.0 if ideal count is too high).
+
+    :param items: The items to be balanced
+    :param rate: The overall rate necessary
+    :param intensity: How close to the ideal should we go from our current position (0=do not change, 1 go to ideal)
+    :param min_budget: Ensure that we use at least min_budget (in order to keep the overall rate)
+
+    :return: A mapping with the frequency for all items + the amount of items used ( it should allways be at least
+    minimum_consumption if passed)
+    """
+
+    total = get_total(classes)
+    num_classes = len(classes)
+
+    if min_budget is None:
+        # use exactly what we need (default handling when we resize everything)
+        min_budget = total * rate
+
+    assert total >= min_budget
+    ideal = total * rate / num_classes
+
+    used_budget: float = 0.0
+
+    ret_val = []
+    while classes:
+        element = classes.pop()
+        count = element.count
+        if ideal * num_classes < min_budget:
+            # if we keep to our ideal we will not be able to use the minimum budget (readjust our target)
+            ideal = min_budget / num_classes
+        # see what's the difference from our ideal
+        sampled = count * rate
+        delta = ideal - sampled
+        correction = delta * intensity
+        desired_count = sampled + correction
+
+        if desired_count > count:
+            # we need more than we have, the best we can do is give all, i.e. rate = 1.0
             new_sample_rate = 1.0
-            total_budget -= count
+            used = count
         else:
-            # we have enough elements in the current class
-            # we want to only keep budget_per_elements
-            new_sample_rate = budget_per_element_type / count
-            total_budget -= budget_per_element_type
-        ret_val.append(
-            DSElement(
-                id=name,
-                new_sample_rate=new_sample_rate,
-                count=count,
-            )
-        )
+            # we can spend what we want
+            new_sample_rate = desired_count / count
+            used = desired_count
+        ret_val.append(DSElement(id=element.id, count=count, new_sample_rate=new_sample_rate))
+        min_budget -= used
+        used_budget += used
+        num_classes -= 1
+
+    return ret_val, used_budget
+
+
+def get_total(transactions: List[DSElement]) -> float:
+    ret_val = 0.0
+    for elm in transactions:
```

Comment:
```suggestion
    for class in classes:
```

Reply:
definitely NOT, using a reserved name as a variable name is not a good idea in my opinion, although it will not
generate an error in this case.

Reply:
I overlooked this and you are right, I would say that we could then keep `elm`.